File: ITP_KANDY_7-main/frontend/lib/features/account/domain/entities/feedback.dart
/**
 * Domain Entity: UserFeedback.
 * Represents a structured support request or bug report submitted by a user.
 * Rationale: Provides a unified data structure for both authenticated owner feedback
 * and unauthenticated public support tickets, ensuring identity traceability for admin triage.
 */
class UserFeedback {
  // --- Identity & Ownership ---
  final String id; // Protocol: Unique identifier (MongoDB _id)
  final String ownerId; // Traceability: Links to the registered Owner (empty for public tickets)
  final String ownerName; // Display: Cached name for high-speed admin listing
  
  // --- Content & Classification ---
  final String category; // Taxonomy: Classification (e.g., 'Error', 'Feature')
  final String message; // Payload: The descriptive body of the support request
  final DateTime createdAt; // Metadata: Timestamp for prioritization
  
  // --- Public Account Recovery Details ---
  final String? contactInfo; // Recovery: Phone/Email for unauthenticated users
  final String? claimedShopName; // Recovery: Used by public users to identify their locked store
  final bool isVerified; // Security: OTP status for identity-sensitive recovery requests

  UserFeedback({
    required this.id,
    required this.ownerId,
    required this.ownerName,
    required this.category,
    required this.message,
    required this.createdAt,
    this.contactInfo,
    this.claimedShopName,
    this.isVerified = false,
  });

  /**
   * Logic: Secure Deserialization.
   * Rationale: Handles multi-tenant identity mapping and ensures MongoDB compatibility.
   */
  factory UserFeedback.fromJson(Map<String, dynamic> json) {
    return UserFeedback(
      // Strategy: Handle both standard 'id' and MongoDB '_id' variants.
      id: json['id'] ?? (json['_id'] ?? ''),
      ownerId: json['ownerId'] ?? '',
      ownerName: json['ownerName'] ?? '',
      category: json['category'] ?? '',
      message: json['message'] ?? '',
      contactInfo: json['contactInfo'],
      claimedShopName: json['claimedShopName'],
      isVerified: json['isVerified'] ?? false,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt']).toLocal()
          : DateTime.now(),
    );
  }

  /**
   * Logic: Administrative Serialization.
   * Rationale: Encapsulates feedback payload for transmission to the triage backend.
   */
  Map<String, dynamic> toJson() {
    return {
      'ownerName': ownerName, // Cached: Speeds up admin dashboard rendering
      'category': category,
      'message': message,
      'contactInfo': contactInfo,
      'claimedShopName': claimedShopName,
      'isVerified': isVerified, // Trace: Proves the claimant verified their phone via OTP
    };
  }
}

File: ITP_KANDY_7-main/backend/src/domain/entities/Feedback.js
/**
 * Domain Layer: Feedback Entity.
 * Represents a merchant's communication with the system administrators.
 * Encapsulates support requests, bug reports, and account recovery documentation.
 */
class Feedback {
  /**
   * Logic: Entity Initialization.
   * Maps heterogeneous support data into a structured business object for administrative review.
   */
  constructor({
    id, // Unique identity (UUID)
    ownerId, // Link to the merchant (if authenticated)
    ownerName, // Snapshot of the sender's name for administrative clarity
    category, // Logic: 'Feedback', 'Error', 'Improvement', 'Account Recovery'
    message, // The core content of the submission
    createdAt, // Audit: When the feedback was transmitted
    contactInfo, // Fallback communication channel for unauthenticated users
    claimedShopName, // Identity proof for manual account recovery requests
    isVerified, // Security: Whether the sender passed an OTP challenge
  }) {
    this.id = id;
    this.ownerId = ownerId;
    this.ownerName = ownerName || 'Unknown User'; // Privacy: Default if name is missing
    this.contactInfo = contactInfo || null;
    this.claimedShopName = claimedShopName || null;
    this.isVerified = isVerified ?? false; // Security: Guaranteed boolean state
    this.category = category;
    this.message = message;
    this.createdAt = createdAt || new Date().toISOString(); // Audit timeline
  }

  /**
   * Logic: Data Serialization.
   * Converts the entity into a JSON object for storage in MongoDB or display in the Admin Panel.
   */
  toJSON() {
    return {
      id: this.id,
      ownerId: this.ownerId,
      ownerName: this.ownerName,
      category: this.category,
      message: this.message,
      createdAt: this.createdAt,
      contactInfo: this.contactInfo,
      claimedShopName: this.claimedShopName,
      isVerified: this.isVerified,
    };
  }
}

// Module Export: Business logic representation of a support or feedback event.
module.exports = Feedback;
